drivers/nec/display/display.cr

```diff
@@ -0,0 +1,313 @@
+    logger.debug { "Polling, power = #{current_power}" }
+
+    if current_power
+      mute_status
+      volume_status
+      video_input
+      audio_input
+    end
+  end
+
+  # LCD Response code
+  def received(data, task)
+    logger.debug { "task is #{task.try &.name}" }
+
+    ascii_string = String.new(data)
+    # Check for valid response
+    if !check_checksum(data)
+      return task.try &.retry("-- NEC LCD, invalid response was: #{ascii_string}")
+    end
+
+    logger.debug { "NEC LCD responded with ascii_string #{ascii_string}" }
+
+    command = MsgType.from_value(data[4])
+
+    case command # Check the MsgType (B, D or F)
+    when .command_reply?
+      # Power on and off
+      if ascii_string[10..15] == "C203D6" # Means power comamnd
+        # 8..9 == "00" means no error
+        if ascii_string[8..9] == "00"
+          self[:power] = ascii_string[11] == '1'
+        else
+          return task.try &.abort("-- NEC LCD, command failed: #{command}\n-- NEC LCD, response was: #{ascii_string}")
+        end
+      elsif ascii_string[12..13] == "D6" # Power status response
+        # 10..11 == "00" means no error
+        if ascii_string[10..11] == "00"
+          self[:power] = ascii_string[23] == '1'
+        else
+          return task.try &.abort("-- NEC LCD, command failed: #{command}\n-- NEC LCD, response was: #{ascii_string}")
+        end
+      end
+    when .get_parameter_reply?, .set_parameter_reply?
+      if ascii_string[8..9] == "00"
+        parse_response(ascii_string)
+      elsif ascii_string[8..9] == "BE"    # Wait response
+        return task.try &.retry("-- NEC LCD, response was a wait command")
+      else
+        return task.try &.abort("-- NEC LCD, command failed: #{command}\n-- NEC LCD, response was: #{ascii_string}")
+      end
+    end
+
+    task.try &.success
+  end
+
+  private def parse_response(data : String)
+    # 14..15 == type (we don't care)
+    value = data[20..23].to_i(16)
+    command = Command.from_value(data[10..13].to_i(16))
+
+    logger.debug { "command is #{data[10..13]}" }
+    logger.debug { command }
+    logger.debug { "value is #{data[20..23]}" }
+    logger.debug { value }
+
+    case command
+    when .video_input?
+      self[:input] = Input.from_value(value)
+    when .audio_input?
+      self[:audio] = Audio.from_value(value)
+    when .volume_status?
+      self[:volume] = value
+    when .brightness_status?
+      self[:brightness] = value
+    when .contrast_status?
+      self[:contrast] = value
+    when .mute_status?
+      self[:audio_mute] = value == 1
+      self[:volume] = 0 if value == 1
+    when .auto_setup?
+      # auto_setup
+      # nothing needed to do here (we are delaying the next command by 4 seconds)
+    else
+      logger.info { "-- NEC LCD, unknown response: #{data[10..13]}" }
+      logger.info { "-- NEC LCD, full response was: #{data}" }
+    end
+  end
+
+  # Types of messages sent to and from the LCD
+  enum MsgType
```

Comment:
You can specify int subtypes for an enums value. In this case:

```crystal
enum MsgType : UInt8
  ...
end
```

which will remove some of the casting your needing to do with this elsewhere.